### PR TITLE
add test cases for GradScaler on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109994

### CODE REVIEW

**Comment by CaoE** [comment]
@jgong5 Could you please help review this PR ? Thank you.

**Review by jgong5 — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: test/test_torch.py [hunk 1]
```diff
@@ -5707,6 +5708,453 @@ def test_grad_scaling_update_scale(self, device, dtype):
...
+        scaler = GradScaler(init_scale=2.0)
+
+        with torch.autocast(device_type=device.type, dtype=torch.half):
+            y = m(x)
+            loss = y.mean()
+        scaler.scale(loss).backward()
+        with self.assertWarns(FutureWarning):
+            scaler.step(o1)
+        scaler.step(o2)
+        scaler.update()
```
Comment (ezyang) [inline test/test_torch.py:6156]: You should seriously just consider making a dedicated test file for this though

Comment (CaoE) [reply test/test_torch.py:6156]: Do you mean making a dedicated test file like test_grad_scaler.py for all GradScaler related tests above ? It's okay for me. I will revise this PR.

File: test/test_torch.py [hunk 2]
```diff
@@ -5707,6 +5708,453 @@ def test_grad_scaling_update_scale(self, device, dtype):
         self.assertEqual(growth_tracker, 0)
         self.assertEqual(scale, 2.0)
 
+    @skipIfTorchDynamo("Failed running call_function for sparse_coo_tensor. See https://github.com/pytorch/pytorch/issues/118856")
+    @onlyNativeDeviceTypes
+    @dtypes(torch.float)
+    def test_grad_scaling_unscale_sparse(self, device, dtype):
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_torch.py:5714]: No, only the device change.
The simple code would also fail：
`torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in method sparse_coo_tensor of type object at 0x7f105afdb9a0>(*(FakeTensor(..., size=(2, 3), dtype=torch.int64), FakeTensor(..., size=(3,)), (2, 3)), **{'device': 'cpu', 'dtype': torch.float32}):
The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.`
```
@onlyNativeDeviceTypes
@dtypes(torch.float)
def test_grad_scaling_unscale_sparse(self, device, dtype):
    i = torch.tensor([[0, 1, 1],
                      [2, 0, 2]], device="cpu", dtype=torch.int64)
    v = torch.tensor([16., 32., 64.], device="cpu", dtype=torch.float)
    s = torch.sparse_coo_tensor(i, v, torch.Size([2, 3]), device="cpu", dtype=torch.float)
```

**Pull request closed without merging**